tests/firecrackerdrivertest.c

```diff
@@ -0,0 +1,311 @@
+{
+    int ret = 0;
+
+    virConnectPtr conn = virConnectOpen("fc:///system");
+
+    if (conn == NULL)
+        ret = -1;
+
+    if (virConnectClose(conn) < 0)
+        ret = -1;
+
+    return ret;
+}
+
+static virConnectPtr test_conn;
+
+static int
+testInitiateConnection(const void *opaque G_GNUC_UNUSED)
+{
+    test_conn = virConnectOpen("fc:///system");
+
+    if (test_conn == NULL)
+        return -1;
+
+    return 0;
+}
+
+static int
+testDefineDomain(const void *opaque G_GNUC_UNUSED)
+{
+    g_autofree const char *path = g_strdup_printf("%s/firecrackerdriverdata/%s.xml",
+                                      abs_srcdir,
+                                      "test_domain");
+
+    char *xml = NULL;
+    if (virFileReadAll(path, 10 * 1024 * 1024, &xml) < 0) {
+        return -1;
+    }
+
+    VIR_DEBUG("xml read: %s", xml);
+
+    if (virDomainDefineXML(test_conn, xml) == NULL) {
+        return -1;
+    }
+
+    return 0;
+}
+
+static int
+testCreateDomain(const void *opaque G_GNUC_UNUSED)
+{
+    virDomainPtr dom = NULL;
+    int state = VIR_DOMAIN_NOSTATE;
+    int max_retries = 100;
+
+    if ((dom = virDomainLookupByName(test_conn, "firecracker_domain")) == NULL) {
+        return -1;
+    }
+
+    if (virDomainCreate(dom) < 0) {
+        return -1;
+    }
+
+    // Usually firecracker is set into running state even if the vm isn't
+    // yet ready to receive other commands like 'SendCtrlAltDel'
+    g_usleep(2500 * 1000);
+
+    do {
+        if (virDomainGetState(dom, &state, NULL, 0) < 0) {
+            VIR_DEBUG("Error getting state");
+            return -1;
+        }
+
+        if (max_retries <= 0)
+            return -1;
+
+        --max_retries;
+    }
+    while (state != VIR_DOMAIN_RUNNING);
+
+    return 0;
+}
+
+static int
+testShutdownDomain(const void *opaque G_GNUC_UNUSED)
+{
+    virDomainPtr dom = NULL;
+    int state = VIR_DOMAIN_NOSTATE;
+    int max_retries = 500;
+
+    if ((dom = virDomainLookupByName(test_conn, "firecracker_domain")) == NULL) {
+        return -1;
+    }
+
+    if (virDomainShutdown(dom) < 0) {
+        return -1;
+    }
+
+    do {
+        g_usleep(50 * 1000);
+
+        if (virDomainGetState(dom, &state, NULL, 0) < 0) {
+            VIR_DEBUG("Error getting state");
+            return -1;
+        }
+
+        if (max_retries <= 0)
+            return -1;
+
+        --max_retries;
+    }
+    while (state != VIR_DOMAIN_SHUTOFF);
+
+    return 0;
+}
+
+static int
+testUndefineDomain(const void *opaque G_GNUC_UNUSED)
+{
+    virDomainPtr dom = NULL;
+
+    if ((dom = virDomainLookupByName(test_conn, "firecracker_domain")) == NULL) {
+        return -1;
+    }
+
+    if (virDomainUndefine(dom) < 0) {
+        return -1;
+    }
+
+    return 0;
+}
+
+static int
+testCloseConnection(const void *opaque G_GNUC_UNUSED)
+{
+    if (virConnectClose(test_conn) < 0) {
+        return -1;
+    }
+
+    return 0;
+}
+
+# define DO_SUBTEST(TESTFUNC, arg) \
+    do { \
```

Comment:
why do we have a loop with only 1 iteration?

Reply:
This is the "common" way to have multi line defines that run multiple instructions. Such that we have readable code